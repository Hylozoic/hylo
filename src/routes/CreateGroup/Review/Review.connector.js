import { get } from 'lodash/fp'
import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import getMe from 'store/selectors/getMe'
import updateUserSettings from 'store/actions/updateUserSettings'
import { addGroupName, addGroupDomain, fetchGroupExists } from '../CreateGroup.store'
import { createGroup, getParents } from './Review.store'

export function mapStateToProps (state, props) {
  const groupParentIds = get('parentIds', state.CreateGroup)
  const parentGroups = getParents(state, { parentIds: groupParentIds })

  return {
    currentUser: getMe(state),
    groupDomain: get('domain', state.CreateGroup),
    groupName: get('name', state.CreateGroup),
    groupPrivacy: get('privacy', state.CreateGroup),
    groupParentIds,
    groupDomainExists: get('domainExists', state.CreateGroup),
    groupParentNames: parentGroups.map(g => get('name', g))
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    goToGroup: (groupDomain) => dispatch(push(`/g/${groupDomain}`)),
    updateUserSettings: (changes) => dispatch(updateUserSettings(changes)),
    clearNameFromCreateGroup: () => dispatch(addGroupName(null)),
    clearDomainFromCreateGroup: () => dispatch(addGroupDomain(null)),
    createGroup: (name, slug, parentIds) => dispatch(createGroup(name, slug, parentIds)),
    goToPrivacyStep: () => dispatch(push('/create-group/privacy')),
    goHome: () => dispatch(push('/')),
    fetchGroupExists: (slug) => dispatch(fetchGroupExists(slug))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
