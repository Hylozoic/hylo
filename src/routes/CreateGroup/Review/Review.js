import React, { Component, createRef } from 'react'
import '../CreateGroup.scss'
import ModalSidebar from 'components/ModalSidebar'
import TextInput from 'components/TextInput'
import { hyloNameWhiteBackground, groovingAxolotl } from 'util/assets'
import { bgImageStyle } from 'util/index'
import ModalFooter from 'components/ModalFooter'
import { find } from 'lodash'
import { get } from 'lodash/fp'
import {
  slugValidatorRegex,
  formatDomainWithUrl,
  removeUrlFromDomain,
  invalidSlugMessage
} from '../util'

export default class Review extends Component {
  constructor () {
    super()
    const fields = ['groupName', 'groupDomain']
    this.inputRefs = fields.reduce((acc, name) => {
      acc[name] = createRef()
      return acc
    }, {})
    this.state = {
      edits: fields.reduce((acc, name) => {
        acc[name] = null
        return acc
      }, {}),
      readOnly: fields.reduce((acc, name) => {
        acc[name] = true
        return acc
      }, {})
    }
  }

  editHandler = (name) => {
    this.setState({
      readOnly: {
        ...this.state.readOnly,
        [name]: false
      }
    })
    this.inputRefs[name].current.select()
  }

  handleInputChange = (event, name) => {
    let value = event.target.value
    if (name === 'groupDomain') {
      value = removeUrlFromDomain(value)
      if (value !== '') {
        this.props.fetchGroupExists(value)
      }
    }
    this.setState({
      edits: {
        ...this.state.edits,
        [name]: value,
        changed: true
      }
    })
  }

  onEnter = event => {
    if (event.key === 'Enter') {
      this.errorCheckAndSubmit()
    }
  }

  submit = () => {
    const { groupName, groupDomain, groupNetworkId } = this.state.edits
    this.props.createGroup(
      // groupPrivacy,
      groupName,
      groupDomain,
      groupNetworkId
    )
      .then(({ error }) => {
        if (error) {
          this.setState({
            error: 'There was an error, please try again.'
          })
        } else {
          this.props.goToGroup(`${groupDomain}`)
        }
      })
  }

  errorCheckAndSubmit = () => {
    const { groupName, groupDomain } = this.state.edits
    if (groupName === '' || groupDomain === '') {
      this.setState({
        error: 'Please fill in each field.'
      })
    } else if (this.props.groupDomainExists) {
      this.setState({
        error: 'This URL already exists. Try another.'
      })
    } else if (!slugValidatorRegex.test(removeUrlFromDomain(groupDomain))) {
      formatDomainWithUrl(groupDomain)
      this.setState({
        error: invalidSlugMessage
      })
    } else {
      this.submit()
    }
  }

  componentWillMount = () => {
    const { groupPrivacy } = this.props
    const privacyOption = find(privacyOptions, { label: groupPrivacy })
    const selectedGroupPrivacy = get('label', privacyOption) // set to Private by default
    this.setState({
      edits: {
        groupName: get('groupName', this.props) || '',
        groupDomain: get('groupDomain', this.props) || '',
        groupNetworkId: get('groupNetworkId', this.props) || '',
        groupPrivacy: selectedGroupPrivacy
      }
    })
  }

  formatDomainWithUrl (groupDomain) {
    if (!groupDomain) return null
    let formattedDomain = groupDomain.replace('hylo.com/g/', '').replace('hylo.com/g', '')
    if (formattedDomain !== '') {
      formattedDomain = 'hylo.com/g/' + formattedDomain
    }
    return formattedDomain
  }

  removeUrlFromDomain (groupDomain) {
    return groupDomain.replace('hylo.com/g/', '')
  }

  render () {
    const { networkName } = this.props

    return <div styleName='flex-wrapper'>
      <ModalSidebar
        imageUrl={groovingAxolotl}
        onClick={this.props.goHome}
        theme={sidebarTheme}
        header='Everything looking good?'
        body='You can always come back and change your details at any time'
      />
      <div styleName='panel'>
        <div>
          <span styleName='step-count'>STEP 3/3</span>
        </div>
        <div styleName='center'>
          <div styleName='logo center' style={bgImageStyle(hyloNameWhiteBackground)} />
        </div>
        <div styleName='center-review'>
          <ReviewTextInput
            label={'Group Name'}
            value={this.state.edits.groupName || ''}
            readOnly={this.state.readOnly.groupName}
            editHandler={() => this.editHandler('groupName')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'groupName')}
            inputRef={this.inputRefs['groupName']}
          />
          <ReviewTextInput
            label={'URL'}
            value={formatDomainWithUrl(this.state.edits.groupDomain) || ''}
            readOnly={this.state.readOnly.groupDomain}
            editHandler={() => this.editHandler('groupDomain')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'groupDomain')}
            inputRef={this.inputRefs['groupDomain']}
          />
          {networkName && <ReviewTextInput
            label={'Network'}
            value={networkName}
            readOnly
          />}
          {/* <ReviewTextInput */}
          {/* label={'Privacy'} */}
          {/* value={this.state.edits.groupPrivacy} */}
          {/* onEnter={this.onEnter} */}
          {/* editHandler={() => this.props.goToPrivacyStep()} */}
          {/* /> */}

        </div>
        { this.state.error && <span styleName='review-arrow-up' /> }
        { this.state.error && <span styleName='review-error'>{this.state.error}</span>}
      </div>
      <ModalFooter
        submit={this.errorCheckAndSubmit}
        previous={this.previous}
        showPrevious={false}
        continueText={'Finish Up'}
      />
    </div>
  }
}

export function ReviewTextInput ({ label, value, editHandler, onChange, readOnly = true, inputRef }) {
  return <div styleName='review-input-text-row'>
    <div styleName='review-input-text-label'>
      <span>{label}</span>
    </div>
    <div styleName='review-input-text'>
      <TextInput
        type='text'
        name='group-name'
        value={value}
        theme={inputTheme}
        readOnly={readOnly}
        noClearButton
        onChange={onChange}
        inputRef={inputRef}
      />
    </div>
    {editHandler && <div styleName='review-input-edit'>
      <span styleName='edit-button' onClick={editHandler}>Edit</span>
    </div>}
  </div>
}

const inputTheme = {
  inputStyle: 'modal-input partial',
  wrapperStyle: 'center'
}

const sidebarTheme = {
  sidebarHeader: 'sidebar-header-full-page',
  sidebarText: 'gray-text sidebar-text-full-page'
}

const privacyOptions = [
  { id: '0', label: 'public' },
  { id: '1', label: 'private' },
  { id: '2', label: 'unlisted' }
]
