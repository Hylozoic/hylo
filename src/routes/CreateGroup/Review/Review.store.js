import orm from 'store/models'
import { createSelector as ormCreateSelector } from 'redux-orm'
import { AnalyticsEvents } from 'hylo-utils/constants'

export const MODULE_NAME = `Review`
export const CREATE_GROUP = `${MODULE_NAME}/CREATE_GROUP`

export function createGroup (name, slug, parentIds) {
  const data = {
    name,
    slug
  }

  if (parentIds) {
    data.parentIds = parentIds
  }

  return {
    type: CREATE_GROUP,
    graphql: {
      query: `mutation ($data: GroupInput) {
        createGroup(data: $data) {
          id
          hasModeratorRole
          group {
            id
            name
            slug
            parentGroups {
              items {
                id
              }
            }
          }
        }
      }
      `,
      variables: {
        data
      }
    },
    meta: {
      extractModel: 'Membership',
      ...data,
      analytics: AnalyticsEvents.GROUP_CREATED
    }
  }
}

export const getParents = ormCreateSelector(
  orm,
  (state, { parentIds }) => parentIds,
  (session, parentIds) => {
    return parentIds && parentIds.length > 0 ? session.Group.filter(g => parentIds.includes(g.id)).toModelArray() : []
  }
)
