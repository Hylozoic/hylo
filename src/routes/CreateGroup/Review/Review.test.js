import React from 'react'
import { shallow } from 'enzyme'
import Review, { ReviewTextInput } from './Review'

describe('Review', () => {
  it('renders correctly', () => {
    const groupName = 'groupName'
    const groupDomain = 'groupDomain'
    const groupPrivacy = 'groupPrivacy'

    const wrapper = shallow(<Review
      goToNextStep={jest.fn()}
      goToPrivacyStep={jest.fn()}
      currentUser={{}}
      fetchMySkills={jest.fn()}
      groupName={groupName}
      groupDomain={groupDomain}
      groupPrivacy={groupPrivacy}
      goHome={jest.fn()}
    />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('ReviewTextInput', () => {
  it('renders correctly', () => {
    const label = 'label'
    const value = 'value'
    const readOnly = false

    const wrapper = shallow(<ReviewTextInput
      label={label}
      value={value}
      readOnly={readOnly}
      editHandler={jest.fn()}
      onChange={jest.fn()}
    />)
    expect(wrapper).toMatchSnapshot()
  })
})
