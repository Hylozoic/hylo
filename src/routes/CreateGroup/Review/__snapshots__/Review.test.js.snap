// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review renders correctly 1`] = `
<div
  data-stylename="flex-wrapper"
>
  <ModalSidebar
    body="You can always come back and change your details at any time"
    header="Everything looking good?"
    imageUrl="/grooving-axolotl.png"
    onClick={[MockFunction]}
    theme={
      Object {
        "sidebarHeader": "sidebar-header-full-page",
        "sidebarText": "gray-text sidebar-text-full-page",
      }
    }
  />
  <div
    data-stylename="panel"
  >
    <div>
      <span
        data-stylename="step-count"
      >
        STEP 3/3
      </span>
    </div>
    <div
      data-stylename="center"
    >
      <div
        data-stylename="logo center"
        style={
          Object {
            "backgroundImage": "url(/logo-white.svg)",
          }
        }
      />
    </div>
    <div
      data-stylename="center-review"
    >
      <ReviewTextInput
        editHandler={[Function]}
        inputRef={
          Object {
            "current": null,
          }
        }
        label="Group Name"
        onChange={[Function]}
        onEnter={[Function]}
        readOnly={true}
        value="groupName"
      />
      <ReviewTextInput
        editHandler={[Function]}
        inputRef={
          Object {
            "current": null,
          }
        }
        label="URL"
        onChange={[Function]}
        onEnter={[Function]}
        readOnly={true}
        value="hylo.com/c/groupDomain"
      />
    </div>
  </div>
  <ModalFooter
    continueText="Finish Up"
    showPrevious={false}
    submit={[Function]}
  />
</div>
`;

exports[`ReviewTextInput renders correctly 1`] = `
<div
  data-stylename="review-input-text-row"
>
  <div
    data-stylename="review-input-text-label"
  >
    <span>
      label
    </span>
  </div>
  <div
    data-stylename="review-input-text"
  >
    <TextInput
      name="group-name"
      noClearButton={true}
      onChange={[MockFunction]}
      readOnly={false}
      theme={
        Object {
          "inputStyle": "modal-input partial",
          "wrapperStyle": "center",
        }
      }
      type="text"
      value="value"
    />
  </div>
  <div
    data-stylename="review-input-edit"
  >
    <span
      data-stylename="edit-button"
      onClick={[MockFunction]}
    >
      Edit
    </span>
  </div>
</div>
`;
